src/dbnode/persist/fs/persist_manager.go

```diff
@@ -115,22 +115,106 @@ type dataPersistManager struct {
 	snapshotID uuid.UUID
 }
 
-type indexPersistManager struct {
-	writer        IndexFileSetWriter
-	segmentWriter m3ninxpersist.MutableSegmentFileSetWriter
-
+type singleUseIndexWriterState struct {
 	// identifiers required to know which file to open
 	// after persistence is over
 	fileSetIdentifier FileSetFileIdentifier
 	fileSetType       persist.FileSetType
 
-	// track state of writers
-	writeErr    error
-	initialized bool
+	// track state of writer
+	writeErr error
+}
+
+// Support writing to multiple index blocks/filesets during index persist.
+// This allows us to prepare an index fileset writer per block start.
+type singleUseIndexWriter struct {
+	// back-ref to the index persist manager so we can share resources there
+	manager *indexPersistManager
+	writer  IndexFileSetWriter
+
+	state singleUseIndexWriterState
+}
+
+func (s *singleUseIndexWriter) persistIndex(builder segment.Builder) error {
+	// Lock the index persist manager as we're sharing the segment builder as a resource.
+	s.manager.Lock()
+	defer s.manager.Unlock()
+
+	markError := func(err error) {
+		s.state.writeErr = err
+	}
+	if err := s.state.writeErr; err != nil {
+		return fmt.Errorf("encountered error: %w, skipping further attempts to persist data", err)
+	}
+
+	if err := s.manager.segmentWriter.Reset(builder); err != nil {
+		markError(err)
+		return err
+	}
+
+	if err := s.writer.WriteSegmentFileSet(s.manager.segmentWriter); err != nil {
+		markError(err)
+		return err
+	}
+
+	return nil
+}
+
+func (s *singleUseIndexWriter) closeIndex() ([]segment.Segment, error) {
+	s.manager.Lock()
+	defer s.manager.Unlock()
+
+	// This writer will be thrown away after we're done persisting.
+	defer func() {
+		s.state = singleUseIndexWriterState{fileSetType: -1}
+		s.manager = nil
+		s.writer = nil
+	}()
```

Comment:
nit: If this is the case honestly we could probably just avoid even doing anything at the end here.

Reply:
Ah yea, that makes sense, I think we should keep this for hygiene. I suppose I'm being overly defensive here haha.